### PR TITLE
Optimize Pool Updating

### PR DESCRIPTION
**Pull request by timjrobinson**
- In update pools compare all changeable pool fields to what they were previously and only update the pool if some have changed.
- Add list of static schema fields that shouldn't be updated because they never change.
- In Decorate Pools don't update these static fields.

### CODE REVIEW

File: src/lambdas/update-pools.ts [hunk 1]
```diff
@@ -19,10 +19,12 @@ export const handler = async (): Promise<any> => {
     const poolsFromChain = await fetchPoolsFromChain(chainId);
     log(`Sanitizing ${poolsFromChain.length} pools`);
     const pools = sanitizePools(poolsFromChain);
-    log(`Saving ${pools.length} pools for chain ${chainId} to database`);
-    await updatePools(pools);
+    const currentPools = await getPools(chainId);
+    const changedPools = getChangedPools(pools, currentPools);
+    log(`Saving ${changedPools.length} pools for chain ${chainId} to database`);
+    await updatePools(changedPools);
```
Comment (alter-eggo) [inline src/lambdas/update-pools.ts:25]: 🤔 do we really need to wait here for updating pools in dynamoDb?
maybe `void updatePools(changedPools);`

File: src/utils.ts [hunk 2]
```diff
@@ -219,3 +223,54 @@ export function formatPrice(price: Price): Price {
...
+ * of all pools that have changed in newPools
+ */
+
+export function getChangedPools(newPools: Pool[], currentPools: Pool[]) {
+  const currentPoolsMap = Object.fromEntries(currentPools.map((pool)=> {
+    return [pool.id, pool];
+  }));
+
+  return newPools.filter((pool) => {
+    if (isSame(pool, currentPoolsMap[pool.id])) {
```
Comment (alter-eggo) [inline src/utils.ts:239]: return `!isSame(pool, currentPoolsMap[pool.id])`

File: src/utils.ts [hunk 3]
```diff
@@ -219,3 +223,54 @@ export function formatPrice(price: Price): Price {
...
+    if (details.static) {
+      return null;
+    }
+    return name;
+  });
+  return compact(nonStaticFields);
+}
+
+export function isSame(newPool: Pool, oldPool: Pool): boolean {
+  if (!oldPool) return false;
```
Comment (alter-eggo) [inline src/utils.ts:257]: if we check oldPool here, so provided type in arguments is wrong.
may be `oldPool?: Pool`

**Review by alter-eggo — CHANGES_REQUESTED** [review]

**Review by garethfuller — COMMENTED** [review]
There seem to be a lot of differences in the pool schema, not sure that's a relevant thing to bring up here? This is a screenshot of the current pool object on the frontend.

Note that:
- `priceRateProviders` has been added and will be required.
- `unwrappedTokens` is redundant.
- We will be deprecating the `onchain` attribute.
Screenshot 2022-11-30 at 14 22 05

File: src/lambdas/update-pools.ts (hunk 1 above)
Comment (timjrobinson) [reply src/lambdas/update-pools.ts:25]: Yea in the update function it processes 25 at a time to not overload the database and takes a while to save them all. Without the wait the lambda would terminate early and not save them all.

File: src/constants/dynamodb.ts [hunk 4]
```diff
@@ -57,34 +57,74 @@ export const TOKENS_TABLE_SCHEMA = {
...
 
-  createTime: { type: 'Int' },
-  swapsCount: { type: 'BigInt' },
-  holdersCount: { type: 'BigInt' },
+  tokens: { type: 'Object', static: false},
+  tokensList: { type: 'Array', static: false },
+  tokenAddresses: { type: 'Array', static: false },
+
+  principalToken: { type: 'String', static: true },
+  baseToken: { type: 'String', static: true },
```
[the remark this answers is not in the document]
Comment (timjrobinson) [reply src/constants/dynamodb.ts:78]: They're for element pools. Will remove.

File: src/constants/dynamodb.ts [hunk 5]
```diff
@@ -57,34 +57,74 @@ export const TOKENS_TABLE_SCHEMA = {
...
 
   // LinearPool only
-  mainIndex: { type: 'Int' },
-  wrappedIndex: { type: 'Int' },
-  lowerTarget: { type: 'BigDecimal' },
-  upperTarget: { type: 'BigDecimal' },
+  mainIndex: { type: 'Int', static: true },
+  wrappedIndex: { type: 'Int', static: true },
+  lowerTarget: { type: 'BigDecimal', static: true },
+  upperTarget: { type: 'BigDecimal', static: true },
```
[the remark this answers is not in the document]
Comment (timjrobinson) [reply src/constants/dynamodb.ts:111]: They are for linear pools. Were recently added to the SDK, not sure what their use is.

**Comment by timjrobinson** [comment]
@garethfuller @alter-eggo please re-review. Should be good to go now.

**Review by alter-eggo — APPROVED** [review]

**Pull request merged**